$OpenBSD: patch-src_iosource_Packet_cc,v 1.8 2021/02/07 16:30:47 ajacoutot Exp $

Prefer DLT_LOOP over DLT_NULL:
DLT_LOOP: 4-byte header with the AF in network byte order
DLT_NULL: 0 hlen and protocol default to AF_INET

Index: src/iosource/Packet.cc
--- src/iosource/Packet.cc.orig
+++ src/iosource/Packet.cc
@@ -81,6 +81,9 @@ int Packet::GetLinkHeaderSize(int link_type)
 	{
 	switch ( link_type ) {
 	case DLT_NULL:
+		return 0;
+
+	case DLT_LOOP:
 		return 4;
 
 	case DLT_EN10MB:
@@ -133,7 +136,13 @@ void Packet::ProcessLayer2()
 	switch ( link_type ) {
 	case DLT_NULL:
 		{
-		int protocol = (pdata[3] << 24) + (pdata[2] << 16) + (pdata[1] << 8) + pdata[0];
+		pdata += GetLinkHeaderSize(link_type);
+		l3_proto = L3_IPV4;
+		break;
+		}
+	case DLT_LOOP:
+		{
+		int protocol = (pdata[0] << 24) + (pdata[1] << 16) + (pdata[2] << 8) + pdata[3];
 		pdata += GetLinkHeaderSize(link_type);
 
 		// From the Wireshark Wiki: "AF_INET6, unfortunately, has
